core/txpool/legacypool/celo_list.go

```diff
@@ -0,0 +1,250 @@
+package legacypool
+
+import (
+	"math"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/txpool"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type celo_list struct {
+	list      *list
+	totalCost map[common.Address]*big.Int
+
+	// Pointer reference to inner list
+	txs *sortedMap
+}
+
+func newCeloList(strict bool) *celo_list {
+	inner_list := newList(strict)
+	return &celo_list{
+		list:      inner_list,
+		totalCost: make(map[common.Address]*big.Int),
+
+		txs: inner_list.txs,
+	}
+}
+
+func (c *celo_list) TotalCostFor(feeCurrency *common.Address) *big.Int {
+	if feeCurrency == nil {
+		return c.list.totalcost
+	}
+	if tc, ok := c.totalCost[*feeCurrency]; ok {
+		return tc
+	}
+	return new(big.Int)
+}
+
+// TotalCost Returns the total cost for transactions with the same fee currency.
+func (c *celo_list) TotalCost(tx *types.Transaction) *big.Int {
+	if !txpool.IsFeeCurrencyTx(tx) {
+		return c.TotalCostFor(nil)
+	}
+	return c.TotalCostFor(tx.FeeCurrency())
+}
+
+func (c *celo_list) addTotalCost(tx *types.Transaction) {
+	if txpool.IsFeeCurrencyTx(tx) {
+		feeCurrency := tx.FeeCurrency()
+		if _, ok := c.totalCost[*feeCurrency]; !ok {
+			c.totalCost[*feeCurrency] = big.NewInt(0)
+		}
+		c.totalCost[*feeCurrency].Add(c.totalCost[*feeCurrency], tx.Cost())
+	} else {
+		c.list.totalcost.Add(c.list.totalcost, tx.Cost())
+	}
+}
+
+func (c *celo_list) subTotalCost(txs types.Transactions) {
+	for _, tx := range txs {
+		if txpool.IsFeeCurrencyTx(tx) {
+			feeCurrency := tx.FeeCurrency()
+			c.totalCost[*feeCurrency].Sub(c.totalCost[*feeCurrency], tx.Cost())
+		} else {
+			c.list.totalcost.Sub(c.list.totalcost, tx.Cost())
+		}
+	}
+}
+
+func (c *celo_list) FilterWhitelisted(st *state.StateDB, all *lookup, fcv txpool.FeeCurrencyValidator) types.Transactions {
+	removed := c.list.txs.Filter(func(tx *types.Transaction) bool {
+		return txpool.IsFeeCurrencyTx(tx) && fcv.IsWhitelisted(st, tx.FeeCurrency())
+	})
+	c.subTotalCost(removed)
+	return removed
+}
+
+func balanceMinusL1Cost(st *state.StateDB, l1Cost *big.Int,
+	feeCurrency *common.Address, addr common.Address,
+	fcv txpool.FeeCurrencyValidator) *big.Int {
+	balance := fcv.Balance(st, addr, feeCurrency)
+	if l1Cost == nil {
+		return balance
+	}
+	currencyL1Cost := fcv.ToCurrencyValue(st, l1Cost, feeCurrency)
```

Comment:
The L1 cost calculation won't work for us like this. But we decided to keep the Optimism behaviour until we know enough about our DA solution to properly handle the DA cost calculation. So that is ok for now.